### PR TITLE
Layer panel

### PR DESCRIPTION
**Pull request by drodenberg**

### CODE REVIEW

**Comment by ntatko** [comment]
I'm not super thrilled about including geojson inside of ol-kit. This is a framework, and should be agnostic of any data applications that it may be used for.

**Comment by drodenberg** [comment]
Oh yeah, don't worry about that. I accidentely made this branch off of the hackathon branch and I'm gonna get rid of all of that. It'll just be LayerPanel code.

**Comment by drodenberg** [comment]
All of the hackathon code should be removed now

File: package.json [hunk 1]
```diff
@@ -53,14 +53,20 @@
     "url": "git@github.com:MonsantoCo/ol-kit.git"
   },
   "dependencies": {
+    "@material-ui/core": "^4.9.12",
+    "@material-ui/icons": "^4.9.1",
+    "@material-ui/styles": "^4.9.10",
```
Comment (drodenberg) [inline package.json:58]: move to dev and peer

File: src/LayerPanel/LayerPanel.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+import LayerPanelBase from 'LayerPanel/LayerPanelBase'
+import LayerPanelLayersPage from 'LayerPanel/LayerPanelLayersPage'
```
Comment (drodenberg) [inline src/LayerPanel/LayerPanel.js:4]: move screenshots to config/static

File: src/LayerPanel/LayerPanelActionExport/index.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  handleMenuClose: PropTypes.func,
+
+  /** An array of openlayers layers */
+  layers: PropTypes.array,
+
+  /** An object of translation key/value pairs */
+  translations: PropTypes.object
+}
+
+export default LayerPanelActionExport
```
Comment (drodenberg) [inline src/LayerPanel/LayerPanelActionExport/index.js:75]: anything that needs translations map or selection can connect to map

File: src/LayerPanel/LayerPanelActionImport/index.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+import MenuItem from '@material-ui/core/MenuItem'
+import { UploadInput } from './styled'
+
```
Comment (drodenberg) [inline src/LayerPanel/LayerPanelActionImport/index.js:5]: make comment annotations. look at popup for reference

File: src/LayerPanel/LayerPanelActions/index.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+LayerPanelActions.propTypes = {
+  /** An bject with key/value pairs for translated strings */
+  translations: PropTypes.object,
+
+  /** An array of components to be displayed inside `LayerPanelMenu` (like `@material-ui/core/MenuItems`) */
+  children: PropTypes.node,
+
+  /** An icon component for the button to open the `LayerPanelMenu` (like `@material-ui/icons`) */
+  icon: PropTypes.node
+}
```
Comment (drodenberg) [inline src/LayerPanel/LayerPanelActions/index.js:59]: make default icon

File: src/LayerPanel/LayerPanelLayersPage/index.js [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+  getVisibleFeaturesForLayer = (layer) => {
+    return (this.isValidVectorLayer(layer))
+      ? layer.getSource().getFeatures().filter(feature => feature.get('_feature_visibility')) : []
+  }
+
+  setVisibilityForAllFeaturesOfLayer = (layer, visibility) => {
+    if (this.isValidVectorLayer(layer)) {
+      layer.getSource().getFeatures().map(feature => {
+        feature.set('_feature_visibility', visibility)
+        feature.setStyle(feature.get('_feature_style'))
```
Comment (drodenberg) [inline src/LayerPanel/LayerPanelLayersPage/index.js:184]: make these `_ol_kit_feature_style`

File: src/LayerPanel/LayerPanelList/index.js [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+  if (addedIndex !== null) {
+    result.splice(addedIndex, 0, layerToAdd)
+  }
+
+  return result
+}
+
+/**
+ * @component
+ * @category vmc
```
Comment (drodenberg) [inline src/LayerPanel/LayerPanelList/index.js:31]: update all these annotations to match popup

File: src/LayerPanel/example.md [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+          <MenuItem onClick={this.setLegendsToOriginalOrder}>Original Order Legends</MenuItem>
+        </LayerPanelActions>
+      } />
+    <LayerPanelContent>
+      <LayerPanelList items={legends} />
+    </LayerPanelContent>
+  </LayerPanelPage>
+</LayerPanel>
+```
+![image](https://github.platforms.engineering/storage/user/2990/files/cf28cb00-86d3-11ea-850b-a33275dd4c1c)
```
Comment (drodenberg) [inline src/LayerPanel/example.md:79]: move these to the config/static file

File: package.json [hunk 9]
```diff
@@ -108,6 +115,9 @@
     "ol": "4.6.5",
     "react": ">=16.8.0",
     "react-dom": ">=16.8.0",
-    "styled-components": ">=5.0.0"
+    "styled-components": ">=5.0.0",
+    "@material-ui/core": ">=4.0.0",
+    "@material-ui/icons": ">=4.9.0",
```
Comment (drodenberg) [inline package.json:120]: change to >=4

File: src/LayerPanel/LayerPanel.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+import { connectToMap } from 'Map'
+
+/**
+ * @component
+ * @category Layerpanel
+ * @since 0.4.0
```
Comment (drodenberg) [inline src/LayerPanel/LayerPanel.js:10]: change to 0.5.0

File: src/LayerPanel/LayerPanel.test.js [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+    // add a feature to that map at a known pixel location
+    const features = [new olFeature(new olPoint([-97.75, 30.265]))]
+
+    const source = new olSourceVector({ features })
+    const vectorLayer = new olLayerVector({ source })
+
+    vectorLayer.set('title', 'My Custom Layer')
+    testMap.addLayer(vectorLayer)
+    testMap.getLayers().getArray()[0].set('title', 'My Basemap')
```
Comment (drodenberg) [inline src/LayerPanel/LayerPanel.test.js:32]: get by `_ol_kit_basemap`

File: src/LayerPanel/LayerPanel.test.js [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+
+    expect(testMap.getLayers().getArray().length).toBe(2)
+
+    await waitFor(() => expect(getByText('My Custom Layer')).toBeInTheDocument())
```
Comment (drodenberg) [inline src/LayerPanel/LayerPanel.test.js:36]: try and get the .not.tobbeindocument to work again

File: src/LayerPanel/LayerPanel.test.js [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+    fireEvent.click(checkboxes[0])
+
+    // make sure the basemap is still visible and that the only layer rendered in the layer panel is not visible
+    expect(testMap.getLayers().getArray()[0].getVisible()).toBe(true)
+    expect(testMap.getLayers().getArray()[1].getVisible()).toBe(true)
+    expect(testMap.getLayers().getArray()[2].getVisible()).toBe(true)
+
+    fireEvent.click(checkboxes[1])
+
+    expect(testMap.getLayers().getArray()[0].getVisible()).toBe(true)
```
Comment (drodenberg) [inline src/LayerPanel/LayerPanel.test.js:81]: comment the indeterminatecheckbox test

File: src/LayerPanel/LayerPanel.test.js [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+    fireEvent.click(getByTestId('LayerPanel.expandLayer'))
+
+    await waitFor(() => expect(getByText('feature 0')).toBeInTheDocument())
+
+    const checkboxes = getAllByTestId('LayerPanel.checkbox')
+
+    fireEvent.click(checkboxes[2])
+
+    expect(getAllByTestId('LayerPanel.indeterminateCheckbox').length).toBe(2)
```
Comment (drodenberg) [inline src/LayerPanel/LayerPanel.test.js:177]: comment this

File: src/LayerPanel/LayerPanel.test.js [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+
+    const source = new olSourceVector({ features })
+    const vectorLayer = new olLayerVector({ source })
+
+    vectorLayer.set('title', 'My Custom Layer')
+    testMap.addLayer(vectorLayer)
+
+    fireEvent.click(getByTestId('LayerPanel.expandLayer'))
+
+    await waitFor(() => expect(getByText('feature 0')).toBeInTheDocument())
```
Comment (drodenberg) [inline src/LayerPanel/LayerPanel.test.js:171]: change feature from being 0 based

File: src/LayerPanel/LayerPanelLayersPage/index.js [hunk 16]
```diff
@@ -0,0 +1,421 @@
...
+                  { features
+                    ? <Collapse in={expandedLayer} timeout='auto' unmountOnExit>
+                      <List component='div' disablePadding style={{ paddingLeft: '36px' }}>
+                        {features.map((feature, i) => {
+                          return (
+                            <ListItem key={i} onDoubleClick={() => handleFeatureDoubleClick(feature)}>
+                              <LayerPanelCheckbox
+                                handleClick={(event) => this.handleFeatureCheckbox(layer, feature, event)}
+                                checkboxState={feature.get('_ol_kit_feature_visibility')} />
+                              <ListItemText inset={false} primary={feature.get('name') || `${translations['_ol_kit.LayerPanelListItem.feature']} ${i}`} />
```
Comment (drodenberg) [inline src/LayerPanel/LayerPanelLayersPage/index.js:347]: add + 1 to the i

**Pull request merged**